Adding the CC2 model energy

AaronDC60 opened the pull request:
## Description

This PR will add the CC2 model energy.

## Todos

Notable points that this PR has either accomplished or will accomplish.

- [x] Add CC2 method to `ccwfn.py`.
- [x] Add test case to compare the CC2 energy to Psi4.

## Status

- [x] Ready to go

bgpeyton commented:
change "CCD" to "CC2" in test name, otherwise LGTM 👍 
ping @lothian

lothian commented on pycc/ccwfn.py:
```diff
@@ -299,6 +299,11 @@ def build_Wmnij(self, o, v, ERI, t1, t2):
+            Wmnij = ERI[o,o,o,o].copy()
+            Wmnij = Wmnij + contract('je,mnie->mnij', t1, ERI[o,o,o,v])
+            Wmnij = Wmnij + contract('ie,mnej->mnij', t1, ERI[o,o,v,o])
+            Wmnij = Wmnij + contract('ijef,mnef->mnij', self.build_tau(t1, t2, fact1=0), ERI[o,o,v,v])
```
Isn't the default `fact1=1.0`?  If so, is this block different from the CCSD block below it?

lothian replied:
I mis-read this in every single instance!  That `1` after `fact` messes with my eyes!

lothian commented on pycc/ccwfn.py:
```diff
@@ -336,6 +341,8 @@ def build_Wmbje(self, o, v, ERI, t1, t2):
         if self.model == 'CCD':
             return
+        elif self.model == 'CC2':
+            return contract('mbef,ijef->mbij', ERI[o,v,v,v], self.build_tau(t1, t2, fact1=0))
```
Same comment here as for Wmnij above.

lothian commented on pycc/ccwfn.py:
```diff
@@ -364,6 +371,22 @@ def r_T2(self, o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi, Wmnij, Wmbej, Wmbje, Zmbi
             r_T2 = r_T2 + contract('imae,mbej->ijab', (t2 - t2.swapaxes(2,3)), Wmbej)
             r_T2 = r_T2 + contract('imae,mbej->ijab', t2, (Wmbej + Wmbje.swapaxes(2,3)))
             r_T2 = r_T2 + contract('mjae,mbie->ijab', t2, Wmbje)
+        elif self.model == 'CC2':
```
I'm not sure whether an entirely new code block for the CC2 case is what we want because there's so much redundancy with CCSD.  Do you think the code is more readable this way?

Please remove any `fact1=1.0` arguments since that's the default.

Also, we should be sure to skip construction of any intermediates that aren't actually needed by CC2, especially Wmbej and Wbmje, which are rather expensive to build.  (See `residuals()`.)

bgpeyton replied:
what do you think is the cleaner option: `if` statements in `residuals()`, or 'if self.model == "CC2": return` in Wmbej and Wbmje (as in Zmbij for the CCD case)?

lothian replied:
I think the latter is cleaner.

AaronDC60 replied:
I did put an entirely new code block for CC2 for the readability, but I can change that if you want.

lothian replied:
No, I'm genuinely not sure which is the better option here.  Let's leave it as it is for now.  It doesn't have to be decided in this PR.

lothian commented:
An important characteristic of the CC2 code should be to avoid all N^6 steps – N^5 max.  That may require modification of some of the current contractions, which were written with CCSD in mind.

AaronDC60 commented:
I changed the code so that Wmbej and Wmbje are no longer build for CC2. I also changed the ```build_Wmnij``` function because in this function only the last line is different between CCSD and CC2.
